tests/clients/test_urllib3.py

```diff
@@ -1,501 +1,575 @@
 # (c) Copyright IBM Corp. 2021
 # (c) Copyright Instana Inc. 2020
 
+import logging
+import sys
 from multiprocessing.pool import ThreadPool
 from time import sleep
-import unittest
+from typing import TYPE_CHECKING
 
-import urllib3
+import pytest
 import requests
-
-import tests.apps.flask_app
-from ..helpers import testenv
+import urllib3
+from instana.instrumentation.urllib3 import (
+    _collect_kvs as collect_kvs,
+    _extract_custom_headers as extract_custom_headers,
+    collect_response,
+)
 from instana.singletons import agent, tracer
 
+import tests.apps.flask_app  # noqa: F401
+from tests.helpers import testenv
+
+if TYPE_CHECKING:
+    from instana.span.span import InstanaSpan
+    from pytest import LogCaptureFixture
+
 
-class TestUrllib3(unittest.TestCase):
-    def setUp(self):
-        """ Clear all spans before a test run """
+class TestUrllib3:
+    @pytest.fixture(autouse=True)
+    def _setup(self) -> None:
+        """Clear all spans before a test run"""
         self.http = urllib3.PoolManager()
-        self.recorder = tracer.recorder
+        self.recorder = tracer.span_processor
         self.recorder.clear_spans()
 
-    def tearDown(self):
-        """ Ensure that allow_exit_as_root has the default value """
+    @pytest.fixture(autouse=True)
+    def _teardown(self) -> None:
+        """Ensure that allow_exit_as_root has the default value"""
         agent.options.allow_exit_as_root = False
```

Comment:
I tried running them and below is the order of their execution
```
tests/clients/test_urllib3.py::TestUrllib3::test_vanilla_requests 
setup
teardown
testcase
```

`yield` keyword or `request.addfinalizer` method is required to indicate the order of setup and teardown methods. Although I read somewhere that `yield` is more preferred.

Reply:
> `yield` keyword or `request.addfinalizer` method is required to indicate the order of setup and teardown methods. Although I read somewhere that `yield` is more preferred.

In that way we can have only one method that performs everything, like:
```
@pytest.fixture(autouse=True)
def _setup(self) -> None:
    # Clear all spans before a test run
    self.http = urllib3.PoolManager()
    self.recorder = tracer.span_processor
    self.recorder.clear_spans()
    yield
    # teardown
    # Ensure that allow_exit_as_root has the default value
    agent.options.allow_exit_as_root = False
```

I will make the change.